### ISSUE TITLE
Static Linking

### ISSUE THREAD

**Issue by wwderw** [Ethosa/nodesnim#35]
I recently found this project.  I like that it seems to follow along the lines of Godot as a Godot user.  I got my own little hello world not really much, not at this point anyway.  I am wanting to make sure that I can compile it correctly.  I seem to be having a time with trying to compile with static linking.

I am running two rigs.  One is KDE Neon and the other is Win 10.  No cross platform tooling, straight native compilation.  Running the same version of Nim on both, 1.4.8 and using nodesnim 0.4.

I am keeping all of the SDL and Freeglut .a files in a folder called lib in my project folder.

I am running the following command on both:
nim  c --dynlibOverride:SDL2 --passL: "-L lib/" main.nim

I seem to be getting compilation errors related to undefined reference of SDL_SetCursor and SDL_CreateSystemCursor with several files within nodesnim.  However, just compiling without dynlibOvveride and passL, everything runs fine on both OSs.  Am I missing something with the linker flag for both of them? Or is statically linking not possible?

Appreciate the project and looking forward to working with it.  

Thanks for the help.

**Comment by Ethosa** [comment]
freeglut has long been removed from dependencies.

and follow it: https://github.com/nim-lang/sdl2#static-linking-sdl2

It's works for me.

**Comment by Ethosa** [comment]
> it pulled in 0.0.6  

For some reason nimble is pulling the wrong version. I have no idea what it might be.


Try it for KDE Neon https://github.com/nim-lang/sdl2#linux

**Comment by wwderw** [comment]
Hmmm, I think this is a function of my knowing just enough to cause damage.  I'm still at the same point with Linux.  Just compiling works, but trying to link statically will compile, but it's the same size binary as before, so I doubt anything extra is getting in there.  Windows passL that is mentioned on the nim-SDL bindings actually throws compiler errors.  Some flags aren't recognized.  I remove those and keep the rest and it compiles, but same as before, just the same size binary.

I tried just moving the SDL.dlls next to the compiled binary and that worked for those, but I was coming up with other .dlls that are missing that aren't apart of what I got from the various SDL libs (image, mixer, main etc).  

What would be the basic --dynlibOveride and --passL for Linux and windows just for the main SDL library itself, just something that I could go from?

I was thinking that the full command would be something like: 

nim c --dynlibOverride:libSDL2 --passL:"-lSDL2 -lSDL2main" main.nim

for --dynlibOverride, I also tried just SDL2 as well with the same results.

While that does compile, it doesn't add anything to the binary.  To get this to work, should I actually be compiling SDL from source to allow linking in such a manner?  I just was using the MS binaries from SDL and my package manager for Neon.

**Comment by Ethosa** [comment]
Do you compile SDL dll libs on Linux by any chance?
https://github.com/Ethosa/nodesnim/blob/master/src/nodesnim/thirdparty/sdl2.nim#L15-L22
![image](https://user-images.githubusercontent.com/49402667/138235304-d6c830c8-27fe-4320-a061-ceebd222c34a.png)

**Comment by wwderw** [comment]
No, I have yet to compile the libs for either platform.  I got the dev libraries for Linux via the package manager and I got the runtime only binaries from SDL website for Windows.  I put the Windows dlls in the .nimble/bin location as well.

I did update yesterday to Nim 1.6.0 (I just noticed that a new version came out 2 days ago).

**Comment by Ethosa** [comment]
I tried compiling with https://stackoverflow.com/questions/69158045/static-link-sdl2-on-linux
and also got this warning.
![image](https://user-images.githubusercontent.com/49402667/138329069-486f98d4-8a09-4d7b-86e8-29c4a6ebf17f.png)

**Comment by wwderw** [comment]
So, just for giggles, I tried to see if I could come up with a bash script to handle the dependencies in a folder called "lib" that would travel with the binary.  

The script that I tried was:
#!/bin/sh

export DIRNAME="$(dirname "$(readlink -f "$0")")"
export SYSTEM_LD_LIBRARY_PATH="$LD_LIBRARY_PATH"


export MAIN_ASSETS_DIR="$DIRNAME/assets/"

cd "$DIRNAME"

export LD_LIBRARY_PATH="$DIRNAME/lib:$LD_LIBRARY_PATH"
"$DIRNAME/bin/main" "$@"

But alas, it was not meant to be.  The program still just kept on telling me that libSDL was not found etc.
